Fix #1681: Updated schema of state in json files.

rt4914 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->


The schema for questions/explorations has been updated on the web codebase which means that the current develop will work correctly but it won't be able to support alpha content. Therefore we need to update the proto, kotlin code and json files to make sure that it follows the latest schema.

There are mainly two major changes:
1. Wherever there are a "choices" (MultipleChoiceInput & ItemSelectionInput), all the options now have a html as well as content_id.
2. The schema of `rule_specs` have been updated and also renamed to `rule_types_to_inputs`

NOTE: All test cases are passing which means that the implementation/changes are correct. Also, the final test can only be done on alpha content.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Commit:
Updated schema

Commit:
Nit fixes

Commit:
ktlint issue fix

BenHenning commented on domain/src/main/java/org/oppia/domain/util/StateRetriever.kt:
```diff
@@ -455,39 +467,44 @@ class StateRetriever @Inject constructor(
 
-  private fun parseJsonStringList(jsonArray: JSONArray): StringList {
-    val list: MutableList<String> = ArrayList()
+  private fun parseSchemaObjectList(jsonArray: JSONArray): SchemaObjectList {
```
This seems misleading--isn't this actually parsing a list of subtitled html schema objects?

rt4914 replied:
The `StringList` has been removed and we had introduced `SchemaObjectList` which is basically `repeated SchemaObject`. So using `SchemaObjectList` we can actually create any type of list and we dont need to create `StringList` separately. 

If we still plan to introduce it, in that case I am not sure what will be the purpose of `SchemaObjectList` and we should remove it then.

Reference: https://gist.github.com/BenHenning/204b87daec6fb0c632be39bfcc7ff85c

BenHenning replied:
Sorry to be clear, my dispute is with the method name actually. 'parseSchemaObjectList' implies it's just parsing an object list, but it's actually parsing a list of subtitled htmls. We should either generalize this to have the caller specify what kind of type the elements of the list will be so that we know how to parse them, or specialize this & its name to be specific to a list of subtitled html (though I'm guessing we actually want the former per my comment about typing above).

rt4914 replied:
Renamed to `parseSubtitledHtmlList`

rt4914 replied:
Resolving this on behalf of Ben

BenHenning commented on domain/src/main/java/org/oppia/domain/util/StateRetriever.kt:
```diff
@@ -455,39 +467,44 @@ class StateRetriever @Inject constructor(
-    val interactionObjectBuilder = InteractionObject.newBuilder()
+  private fun createCustomizationArgValueFromJson(customizationArgValue: Any): SchemaObject {
+    val schemaObjectBuilder = SchemaObject.newBuilder()
     when (customizationArgValue) {
```
Is this the correct way to know what type of schema object is being parsed? My understanding is that the situation in which the customization argument is being parsed should indicate what type of schema object to expect (e.g. in specific interactions). Otherwise, we can't know when to parse custom objects, or if we're interpreting something to be a plaintext string vs. unicode.

rt4914 replied:
Okay, I got your point that the current code cannot differentiate between:
```
"value": {
    "content_id": "ca_placeholder_0",
    "unicode_str": "Sample unicode string"
}
```
```
"value": {
    "content_id": "ca_placeholder_0",
    "html": "Sample html string"
}
```

But the response that I have received by creating questions on localhost has unicode only in this context:

```
"customization_args": {
    "placeholder": {
        "value": {
            "content_id": "ca_placeholder_0",
            "unicode_str": ""
        }
    },
    "rows": {
        "value": 1
    }
}
```

Other than that I have seen only `content_id` and `html` and therefore I am actually not able to understand this clearly.

------------------------------------------------------------------------------------
But even if there is some case where we have something like this:
```
"choices": {
    "value": [{
        "html": "\u003cp\u003e8 + 7 + 6 + 2\u003c/p\u003e",
        "content_id": "ca_choices_0"
    }, {
        "html": "\u003cp\u003e8000 + 700 + 60 + 2\u003c/p\u003e",
        "content_id": "ca_choices_1"
    }]
}
```

Using unicode:
```
"choices": {
    "value": [{
        "unicode_str": "STR1",
        "content_id": "ca_choices_0"
    }, {
        "unicode_str": "STR2",
        "content_id": "ca_choices_1"
    }]
}
```

In above case I think the question that you are raising is how we will differentiate between the two? If so, then can we use something like this:
`if (customizationArgValue.has("unicode_str"))`
to know whether it has unicode string or not?

I am actually not 100% sure if I have understood your comment and therefore adding whatever I was able to understand.

BenHenning replied:
The unicode vs. html was just an example. My point is that there's no typing information in JSON to differentiate between which SchemaObject to choose. The only way to select a schema object type is to already know which type to select. See https://github.com/oppia/oppia/blob/develop/extensions/interactions/DragAndDropSortInput/DragAndDropSortInput.py#L37 as an example. We know from the definition of drag & drop sort input that its customization argument 'choices' will be a list type.

My suggestion is that we update StateRetriever to parse the type of schema object based on what we know interactions will define since that's the most correct way to do it, rather than trying to infer the type based on type-less JSON.

rt4914 replied:
Updated the implementation in state retriever. PTAL
Also, two points:
1. Suggest if we can improvise this code further.
2. I am actually not getting much lead of the table thing that we talked about. What I did was I tried searching for a `HashMap<String, function>`, if something like this is possible and I got no lead in this.

rt4914 replied:
Resolving this on behalf of Ben

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,61 @@ message Interaction {
+  map<string, SchemaObject> customization_args = 7;
+}
+
+// Structure for a customization args
```
More like: ``Structure for schema objects. See <link> for context.``

Where 'link' corresponds to the Oppia backend's definition for schema-based forms.

rt4914 replied:
I have added a link but please check and confirm if this is correct or not.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,61 @@ message Interaction {
+// Structure for a customization args
+message SchemaObject {
+  oneof object_type {
+    bool bool_value = 1;
+    CustomSchemaValue custom_schema_value = 2;
+    DictObject dict_value = 3;
+    float float_value = 4;
+    string normalized_string = 5;
+    int32 signed_int = 6;
+    SchemaObjectList schema_object_list = 7;
+    string unicode_value = 8;
+    UnicodeOrNone unicode_or_none = 9;
+    SubtitledHtml subtitled_html = 10;
+    SubtitledUnicode subtitled_unicode = 11;
+    uint32 non_negative_int = 12;
+    double real = 13;
```
I'm inclined for us to restrict this just to the list of types we actually need to support among all customization arguments of all interactions the Android app currently or is in the process of supporting. We should fail to parse the JSON for other cases.

rt4914 replied:
Removed following:
```
uint32 non_negative_int = 12;
double real = 13;
```

BenHenning commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/DragAndDropSortInteractionViewModel.kt:
```diff
@@ -29,7 +29,13 @@ class DragAndDropSortInteractionViewModel(
   }
   private val choiceStrings: List<String> by lazy {
-    interaction.customizationArgsMap["choices"]?.setOfHtmlString?.htmlList ?: listOf()
+    val choicesList = mutableListOf<String>()
+    val schemaObjectList =
+      interaction.customizationArgsMap["choices"]?.schemaObjectList?.schemaObjectList
+    schemaObjectList?.forEach { schemaObject ->
+      choicesList.add(schemaObject.subtitledHtml.html)
+    }
+    choicesList
```
```suggestion
    interaction.customizationArgsMap["choices"]
        ?.schemaObjectList
        ?.schemaObjectList
        ?.map { schemaObject -> schemaObject.subtitledHtml.html }
        ?: listOf()
```

This is a bit more idiomatic since it doesn't involve mutating a list, and instead computes it step-by-step or early exits to an empty list.

rt4914 replied:
Updated. Thanks.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -25,7 +25,13 @@ class SelectionInteractionViewModel(
 
   private val choiceStrings: List<String> by lazy {
-    interaction.customizationArgsMap["choices"]?.setOfHtmlString?.htmlList ?: listOf()
+    val choicesList = mutableListOf<String>()
+    val schemaObjectList =
+      interaction.customizationArgsMap["choices"]?.schemaObjectList?.schemaObjectList
+    schemaObjectList?.forEach { schemaObject ->
+      choicesList.add(schemaObject.subtitledHtml.html)
+    }
+    choicesList
```
Ditto for this computation.

rt4914 replied:
Updated Thanks.

BenHenning commented on domain/src/main/assets/5NWuolNcwH6e.json:
```diff
@@ -2377,140 +2373,79 @@
-                  "inputs": {
+                  },
+                  {
+                    "x": "5:2:1:6"
```
Why this new one?

rt4914 replied:
This was added by mistake. I have checked on current develop and removed it accordingly.

BenHenning commented on domain/src/main/assets/umPkwp0L1M0-.json:
```diff
@@ -271,19 +265,18 @@
+                  {
                     "f": {
-                      "denominator": 7,
+                      "denominator": 3,
```
Why has this changed?

rt4914 replied:
I have checked it on develop and the denominator should be 7. Thanks for catching this.

BenHenning reviewed (changes requested):
Thanks @rt4914! Just had a few things that I noticed. Also, can you please add schema versions to all of the JSON files so that we know which state schema version each is at now?

rt4914 commented:
> Thanks @rt4914! Just had a few things that I noticed. Also, can you please add schema versions to all of the JSON files so that we know which state schema version each is at now?

I have added `"states_schema_version": 39,` in all explorations. But in all json files that I am downloading from server I amd not getting any information about the schema version and therefore I have names this key from [here](https://github.com/oppia/oppia/blob/fcb44f8cc6e0e00aaa082045cf8b363daa510479/core/domain/exp_domain.py#L570).

For explorations there is a key named `question_state_data_schema_version` which I have updated to `39`

Commit:
Changes justed by Ben

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -133,7 +131,6 @@ message SubtitledUnicode {
 
-
 // Structure to support UnicodeOrNone
 message UnicodeOrNone {
```
Do we still need this structure? Have all of the ones we actually don't need been removed?

rt4914 replied:
No we don't need this anymore.

rt4914 replied:
Resolving this on behalf of Ben

BenHenning reviewed (changes requested):
Thanks @rt4914! Followed up on your comments regarding StateRetriever.

Also, it's a bit interesting to hear that state schema version wasn't included in the downloaded JSON clientside. That's not what I would expect, though technically the frontend doesn't need to know this so it *might* be stripped out. We may need to change that in the future as part of API versioning for Android.

Commit:
Nit changes

Commit:
Merge remote-tracking branch 'upstream/develop' into updated-state-schema

Commit:
Merge remote-tracking branch 'upstream/develop' into updated-state-schema

Commit:
Updated implementation

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,50 @@ message Interaction {
+}
+
+// Structure for unicode and content id.
+message SubtitledUnicode {
```
Doesn't seem used anywhere above--please remove.

rt4914 replied:
Removed this for now but might need to introduce this back in future keeping https://github.com/oppia/oppia/blob/fcb44f8/schema_utils.py#L50 in mind.

Resolving this on behalf of Ben.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,50 @@ message Interaction {
+    ImageWithRegions image_with_regions = 1;
+    SubtitledHtml subtitled_html = 2;
+    SubtitledUnicode subtitled_unicode = 3;
+    ClickOnImage click_on_image = 4;
```
Doesn't appear to be used above--is that correct? If so, please remove.

rt4914 replied:
Removed this for now but might need to introduce this back in future keeping https://github.com/oppia/oppia/blob/fcb44f8/schema_utils.py#L50 in mind.

Resolving this on behalf of Ben.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,50 @@ message Interaction {
+}
+
+// Structure for a dict object
+message DictObject {
```
Not used--remove.

rt4914 replied:
Removed this for now but might need to introduce this back in future keeping https://github.com/oppia/oppia/blob/fcb44f8/schema_utils.py#L50 in mind.

Resolving this on behalf of Ben.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,50 @@ message Interaction {
+  oneof object_type {
+    bool bool_value = 1;
+    CustomSchemaValue custom_schema_value = 2;
+    DictObject dict_value = 3;
+    float float_value = 4;
```
Unused--please remove.

rt4914 replied:
Removed this for now but might need to introduce this back in future keeping https://github.com/oppia/oppia/blob/fcb44f8/schema_utils.py#L50 in mind.

Resolving this on behalf of Ben.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,50 @@ message Interaction {
+    string normalized_string = 5;
+    int32 signed_int = 6;
+    SchemaObjectList schema_object_list = 7;
+    string unicode_value = 8;
```
Not used--please remove.

rt4914 replied:
Removed this for now but might need to introduce this back in future keeping https://github.com/oppia/oppia/blob/fcb44f8/schema_utils.py#L50 in mind.

Resolving this on behalf of Ben.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -85,7 +85,50 @@ message Interaction {
+    SchemaObjectList schema_object_list = 7;
+    string unicode_value = 8;
+    SubtitledHtml subtitled_html = 9;
+    SubtitledUnicode subtitled_unicode = 10;
+    double real = 11;
```
Not used--please remove.

rt4914 replied:
Removed this for now but might need to introduce this back in future keeping https://github.com/oppia/oppia/blob/fcb44f8/schema_utils.py#L50 in mind.

Resolving this on behalf of Ben.

BenHenning reviewed (approved):
Thanks @rt4914! Mostly just have comments to remove unused parts of the proto--feel free to resolve those on your own. This otherwise LGTM.

Commit:
Merge remote-tracking branch 'upstream/develop' into updated-state-schema

Commit:
Removed unused protos
